pkg/server/stream/v3/subscription.go

```diff
@@ -0,0 +1,176 @@
+package stream
+
+const (
+	explicitWildcard = "*"
+)
+
+// Subscription stores the server view of a given type subscription in a stream.
+type Subscription struct {
+	// wildcard indicates if the subscription currently has a wildcard watch.
+	wildcard bool
+
+	// allowLegacyWildcard indicates that the stream never provided any resource
+	// and is de facto wildcard.
+	// As soon as a resource or an explicit subscription to wildcard is provided,
+	// this flag will be set to false
+	allowLegacyWildcard bool
+
+	// subscribedResourceNames provides the resources explicitly requested by the client
+	// This list might be non-empty even when set as wildcard.
+	subscribedResourceNames map[string]struct{}
+
+	// returnedResources contains the resources acknowledged by the client and the acknowledged versions.
+	returnedResources map[string]string
+}
+
+// newSubscription initializes a subscription state.
+func newSubscription(wildcard bool, initialResourceVersions map[string]string) Subscription {
+	state := Subscription{
+		wildcard:                wildcard,
+		allowLegacyWildcard:     wildcard,
+		subscribedResourceNames: map[string]struct{}{},
+		returnedResources:       initialResourceVersions,
+	}
+
+	if initialResourceVersions == nil {
+		state.returnedResources = make(map[string]string)
+	}
+
+	return state
+}
+
+func NewSotwSubscription(subscribed []string) Subscription {
+	sub := newSubscription(len(subscribed) == 0, nil)
```

Comment:
Isn't it possible for the initial set of resources to contain an explicit wildcard?

Reply:
It is, and it would be handled in `SetResourceSubscription` below. 
This code is only as legacy makes the behavior of "nothing" later not wildcard, hence the split behavior here